### PR TITLE
Add barebone Ruby interface for ConcurrentHashMap

### PR DESCRIPTION
**Pull request by wildmaples**
## Change

First step to improve performance of `concurrent-ruby`'s `Concurrent::Map` https://github.com/Shopify/oracle-truffleruby-collab/issues/17

These methods are implemented following the interface in concurrent-ruby's [synchronized_map_backend file](https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent-ruby/concurrent/collection/map/synchronized_map_backend.rb) and following `Concurrent::Map` [API documentation](http://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/Map.html). 

## Background 

Currently TruffleRuby uses the ConcurrentHashMap-written-in-Ruby backend for Concurrent::Map on `concurrent-ruby` which utilizes a Ruby hash. We are creating a backend on the TruffleRuby side so we are able to optimize the mapping ourselves.

In another change, we'll write the change to patch `concurrent-ruby` with this class, and then implement some parts of this class in Java to make it concurrent. 

As Chris mentions in [this comment](https://github.com/Shopify/oracle-truffleruby-collab/issues/17#issuecomment-814933071), `TruffleRuby::ConcurrentHashMap` is named after the [Java class](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html) and not `Concurrent::Hash` from the `concurrent-ruby` gem.  

## Testing

* I've added specs for all of the new methods in `ConcurrentHashMap`.
* I've tested this change on `concurrent-ruby` gem by swapping the `AtomicReferenceMapBackend` with the new `TruffleRuby::ConcurrentHashMap` in [`map.rb`](https://github.com/ruby-concurrency/concurrent-ruby/blob/7b7900a9133a5c31e60fc55133559ae07c934d69/lib/concurrent-ruby/concurrent/map.rb#L20). 

  * See this branch on `concurrent-ruby` with my changes - https://github.com/ruby-concurrency/concurrent-ruby/compare/master...wildmaples:testing-truffleruby-map-backend?expand=1
  * You can run the above changes with this branch's commit (on Truffleruby) to top hat. 

## Note 

As mentioned in my commit message: This current change passes the specs for `concurrent-ruby`'s `Concurrent::Map` specs except for 3: #dup/#clone, updates don't block reads, and atomicity. Atomicity spec is commented out when I tested this change because it stops the entire test process. 

```
Finished in 1 minute 53.88 seconds (files took 5.95 seconds to load)
60 examples, 2 failures

Failed examples:

rspec ./spec/concurrent/map_spec.rb:747 # Concurrent::Map #dup,#clone
rspec ./spec/concurrent/map_spec.rb:256 # Concurrent::Map updates dont block reads
```

Because this PR covers most of the spec cases, I'm hoping to get it merged first. We can continue investigating the remaining failing specs next.

### CODE REVIEW

**Comment by chrisseaton** [comment]
Great - this is in line with what I was thinking in terms of the API and the semantics. What do you want to do next? You could already try consuming this API in a fork of Concurrent Ruby?

**Comment by wildmaples** [comment]
> Great - this is in line with what I was thinking in terms of the API and the semantics. What do you want to do next? You could already try consuming this API in a fork of Concurrent Ruby?

I'm trying it locally right now -- the specs for the methods added are failing so I'm gonna debug it.

**Comment by wildmaples** [comment]
As mentioned in the description, I followed the interface of `SynchronizedMapBackend`. However, after testing `TruffleRuby::ConcurrentHashMap` in `concurrent-ruby`, I found that there are some methods missing. This is because `SynchronizedMapBackend` inherits from `NonConcurrentMapBackend`, which contains some additional methods. Because we want the `TruffleRubyMapBackend` to inherit from our own `ConcurrentHashMap`, we'd need to implement the missing methods. 

Additionally, the [specs for `atomicity` ](https://github.com/ruby-concurrency/concurrent-ruby/blob/7b7900a9133a5c31e60fc55133559ae07c934d69/spec/concurrent/map_spec.rb#L73)fails with `RSpec::Expectations::ExpectationNotMetError` in a strange loop, stopping the test run entirely. I decided to focus getting the missing methods specs to pass and commented out this spec for now.

**Comment by chrisseaton** [comment]
We've got several options for how we implement the backend beyond just inheriting - we also have inclusion and extension, which are ways Ruby lets you use methods from another module, and we can create a delegator, so a class which is a subclass as before but passes other methods to another object.

**Comment by wildmaples** [comment]
Alright this PR is ready for review! I've updated the description. The current CI failure seems to be unrelated to the change.

**Comment by chrisseaton** [comment]
We should document in https://github.com/oracle/truffleruby/blob/master/doc/user/truffleruby-additions.md#truffleruby-methods-and-classes.

File: src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb
Comment (chrisseaton) [inline src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb:5]: `options` isn't used.

File: src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb
Comment (chrisseaton) [inline src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb:136]: I think this method is only called by other methods which already hold this lock? So it shouldn't be needed.

**Review by chrisseaton — CHANGES_REQUESTED** [review]
@eregon, what we're looking to do here is to discuss the API. Maple is also working on the patch to Concurrent Ruby to use it. When both of those are as everyone is happy... then we'll switch it out for an actual concurrent implementation.

So comments on design of the API please - obviously we do already realise it isn't concurrent.

File: src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb
Comment (wildmaples) [reply src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb:5]: The `Map` child class which will inherit this class [uses the options params](https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent-ruby/concurrent/map.rb#L125) so not having it will raise ArgumentError (which happened when I was testing it)

**Review by eregon — COMMENTED** [review]
Could this backend live in concurrent-ruby?
That way we would never have some kind of mismatch between the APIs due to different concurrent-ruby/truffleruby versions.

The only "API" then would be a few Primitives to create and interact with a ConcurrentHashMap (which would use Ruby semantics for comparing keys).
That would also not be a public or documented API, which might be better anyway.

Regarding the inheritance thing, we could inherit from a concurrent-ruby class and just have this as a non-core file. But that again has the issue that the API might evolve over time and needs even tighter coupling between the two.

So I think a minimal set of Primitives to interact with a ConcurrentHashMap is best here, and the backend Ruby class should live in concurrent-ruby.

**Comment by chrisseaton** [comment]
> So I think a minimal set of Primitives to interact with a ConcurrentHashMap is best here,

These methods _are_ what we think is a sensible minimum set of primitives (give or take).

We did ask if you preferred a class or primitives you said you didn't have a preference.

> That would also not be a public or documented API, which might be better anyway.

I disagree I think if a third-party gem is using it it should be a proper API, documented in https://github.com/oracle/truffleruby/blob/master/doc/user/truffleruby-additions.md and https://github.com/oracle/truffleruby/blob/master/doc/contributor/third-party-assumptions.md.

**Comment by eregon** [comment]
OK, this set of methods seems reasonable and basically the same as https://github.com/ruby-concurrency/concurrent-ruby/blob/master/ext/concurrent-ruby/com/concurrent_ruby/ext/JRubyMapBackendLibrary.java
It's so many methods that I thought it would be the entire Concurrent::Map API but actually not that's just the backend API.

File: src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb
Comment (eregon) [inline src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb:8]: You need to implement `initialize_copy`, that's probably what's needed for passing the dup specs.

**Comment by eregon** [comment]
I'm not sure of the value of the current synchronized implementation though, besides just getting started/see if the specs pass.
Could we rather merge when it actually uses a ConcurrentHashMap?
This is not prone to conflicts or anything, so I think there is no need to merge a partial state early.

File: src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb
Comment (eregon) [inline src/main/ruby/truffleruby/core/truffle/concurrent_hash_map.rb:118]: This one is a bit strange because it's not listed in https://www.rubydoc.info/gems/concurrent-ruby/Concurrent/Map
Yes it's also a public method for other backends, so it's fine.

**Comment by chrisseaton** [comment]
Yes I think we converted from draft to get approval for the API in principle - we can convert back to a draft, implement the concurrency, then come back to you.

**Pull request closed without merging**
